src/AstTransforms.cpp

```diff
@@ -222,234 +222,118 @@ bool UniqueAggregationVariablesTransformer::transform(AstTranslationUnit& transl
-//                        toString(*cur), (isNumberType(argTypes[cur])) ? "number" : "symbol"));
-//            }
-//////
-//            program.addClause(std::unique_ptr<AstClause>(newRule));
-//            program.addRelation(std::unique_ptr<AstRelation>(rel));
-//						// now mutate the clause so that we replace the literal with the head __agg_i(x).
-////						struct replaceAggregateLiteral : public AstNodeMapper {
-////							const AstLiteral& literal;
-////							const AstAtom* replacement;
-////							
-////							replaceAggregateLiteral(const AstLiteral& literal, const AstAtom* replacement) 
-////											: literal(literal), replacement(replacement) {}
-////
-////							std::unique_ptr<AstNode> operator()(std::unique_ptr<AstNode> node) const override {
-////								if (auto* lit = dynamic_cast<AstLiteral*>(node.get())) {
-////									if (*lit == literal) {
-////										std::vector<std::unique_ptr<AstArgument>> args;
-////										for (auto arg : replacement->getArguments()) {
-////												args.emplace_back(arg->clone());
-////										}
-////										auto literalReplacement = std::make_unique<AstAtom>(replacement->getQualifiedName(), std::move(args), replacement->getSrcLoc());
-////										return literalReplacement;
-////									}
-////								}
-////								return node;
-////							}
-////						};
-////						replaceAggregateLiteral update(literal, literalClauseHead);
-////						program.apply(update);
-//
-//
-//						//////auto aggAtom =
-//            ////        std::make_unique<AstAtom>(head->getQualifiedName(), std::move(args), head->getSrcLoc());
-//
-//            ////std::vector<std::unique_ptr<AstLiteral>> newBody;
-//            ////newBody.push_back(std::move(aggAtom));
-//            ////const_cast<AstAggregator&>(agg).setBody(std::move(newBody));
-//						//std::vector<std::unique_ptr<AstLiteral>> newClauseBody;
-//						//for (const auto& lit : clause.getBodyLiterals()) {
-//						//		if (lit != &literal) {
-//						//			newClauseBody.push_back(std::move(lit));
-//						//		}
-//						//}
-//            //std::vector<std::unique_ptr<AstArgument>> args;
-//            //for (auto arg : literalClauseHead->getArguments()) {
-//            //    args.emplace_back(arg->clone());
-//            //}
-//
-//						//auto literalReplacement = std::make_unique<AstAtom>(literalClauseHead->getQualifiedName(), std::move(args), literalClauseHead->getSrcLoc());
-//						//newClauseBody.push_back(std::move(literalReplacement));
-//						//const_cast<AstClause&>(clause).setBody(std::move(newClauseBody));
-					});
-				});
-			});
-
-		return changed;
+					}
+					node->apply(*this);
+					assert(node != nullptr);
+					return node;
+				}
+		};
+		std::cout << "before applying update..." << std::endl;
+	 	std::cout << program << std::endl;	
+		replaceAggregate update(aggregate, variable);
+		foundClause->addToBody(std::unique_ptr<AstLiteral>(aggHead->clone()));
+		foundClause->apply(update);
```

Comment:
This is where mainly I don't understand why this doesn't work. And I also obviously have to change the setup to this to store multiple  clause/aggregate pairs. How do you recommend doing this? Otherwise, I'm just going to do it with a set of pairs of Aggregate/Clause raw pointers.

Reply:
You're applying the nodemapper to foundClause, which is actually a clone of the found clause.

Reply:
Ahhhh Thank you Martin!! Silly!!!

Reply:
It took me a while to find that problem, but since it did apply the mapping, yet nothing changed, I was suspicious.

Reply:
Thank you!